src/org/jitsi/impl/neomedia/rtp/translator/RTCPFeedbackMessageSender.java

```diff
@@ -102,87 +113,287 @@ private long getSenderSSRC()
+                sentFIR = true;
+        }
 
-        fir.setSequenceNumber(
-                getNextSequenceNumber((int) senderSSRC, mediaSenderSSRC));
-        return rtpTranslator.writeControlPayload(fir, destination);
+        return sentFIR;
     }
 
     /**
-     * Sends Full Intra Request (FIR) Commands to media senders/sources with
-     * specific synchronization source identifiers (SSRCs) through a specific
-     * <tt>MediaStream</tt>.
+     * Notifies this instance that an RTP packet has been received from a peer
+     * represented by a specific <tt>StreamRTPManagerDesc</tt>.
      *
-     * @param destination the <tt>MediaStream</tt> through which the FIR
-     * commands are to be sent
-     * @param mediaSenderSSRCs the SSRCs of the media senders/sources
-     * @return <tt>true</tt> if a FIR command was sent; otherwise,
-     * <tt>false</tt>
+     * @param streamRTPManager a <tt>StreamRTPManagerDesc</tt> which identifies
+     * the peer from which an RTP packet has been received
+     * @param buf the buffer which contains the bytes of the received RTP or
+     * RTCP packet
+     * @param off the zero-based index in <tt>buf</tt> at which the bytes of the
+     * received RTP or RTCP packet begin
+     * @param len the number of bytes in <tt>buf</tt> beginning at <tt>off</tt>
+     * which represent the received RTP or RTCP packet
      */
-    public boolean sendFIR(MediaStream destination, int[] mediaSenderSSRCs)
+    public void maybeStopRequesting(
+        StreamRTPManagerDesc streamRTPManager,
+        int ssrc,
+        int pt,
+        byte[] buf,
+        int off,
+        int len)
     {
-        boolean sentFIR = false;
+        KeyframeRequester kfRequester = kfRequesters.get(ssrc);
+        if (kfRequester != null)
+        {
+            kfRequester.maybeStopRequesting(streamRTPManager, pt, buf, off, len);
+        }
+    }
 
-        for (int mediaSenderSSRC : mediaSenderSSRCs)
+    /**
+     * The <tt>KeyframeRequester</tt> is responsible for sending FIR requests to
+     * a specific media sender identified by its SSRC.
+     */
+    class KeyframeRequester
+        extends PeriodicProcessible
+    {
+        /**
+         * The media sender SSRC of this <tt>KeyframeRequester</tt>
+         */
+        private final int mediaSenderSSRC;
+
+        /**
+         * The sequence number of the next FIR.
+         */
+        private final AtomicInteger sequenceNumber;
+
+        /**
+         * The number of FIR that are left to be sent before stopping.
+         */
+        private int remainingRetries;
+
+        /**
+         * Ctor.
+         *
+         * @param mediaSenderSSRC
+         */
+        public KeyframeRequester(int mediaSenderSSRC)
         {
-            if (sendFIR(destination, mediaSenderSSRC))
-                sentFIR = true;
+            super(FIR_RETRY_INTERVAL_MS);
+            this.mediaSenderSSRC = mediaSenderSSRC;
+            this.sequenceNumber = new AtomicInteger(0);
```

Comment:
I'm going to make a very moot point here, I admit. But isn't that easier to read if you initialize sequenceNumber above given that it's final and doesn't get initialized with a constructor argument?